### DISCUSSION TITLE
Base Enterprise Deployment Hangs During Federation

### DISCUSSION THREAD

**Discussion by traetimmerman** [Esri/arcgis-powershell-dsc#515] [Q&A] [answered]
I'm working on deploying a single-machine base enterprise deployment on an EC2 instance in AWS. It appears that everything is working properly except the federation step, which hangs and never proceeds nor produces an error. Any thoughts on what might cause the federation to hang?

**Comment by cameronkroeker** [comment]
Hi @traetimmerman,

Could you provide a bit more detail on where its hanging within the federation resource? For example, is it hanging on generating a portal token, is able to generate the token and is hanging on federating, or is it hanging on updating it to be a hosting server?

If possible could you share the module version, enterprise version, dsc logs, and/or json config? Please feel free to scrub sensitive information such as actual machine names, usernames\passwords, etc.

Thanks,
Cameron K.

**Comment by traetimmerman** [comment]
Hey @cameronkroeker,

Thanks for getting back to me. It's failing to generate a token. I let it sit for an hour this morning, and it eventually failed and produced an error. The ArcGIS Enterprise version is 11.2, the module version is v4.2.0, and the configuration file and log files are attached.
[BaseDeployment-SingleMachine.json](https://github.com/Esri/arcgis-powershell-dsc/files/13866840/BaseDeployment-SingleMachine.json)
[dsc_deployment_logs.zip](https://github.com/Esri/arcgis-powershell-dsc/files/13866841/dsc_deployment_logs.zip)

**Comment by cameronkroeker** [comment]
Thanks @traetimmerman for the json config and dsc logs. I think typically its recommended to use ip addresses in AWS rather than the hostnames. However, on the target node can you launch Powershell, as an Administrator, and share the results of this command?

```
Get-FQDN EC2AMAZ-RQ2AG6M
```

I am wondering if it is trying to generate the token via`https://EC2AMAZ-RQ2AG6M:7443/arcgis/sharing/rest/generateToken` rather than `https://EC2AMAZ-RQ2AG6M.local:7443/arcgis/sharing/rest/generateToken`. 

Do either of these url's work?

Thanks,
Cameron K.

**Comment by traetimmerman** [comment]
Hey @cameronkroeker,

That command returns _EC2AMAZ-RQ2AG6M.local_. https://EC2AMAZ-RQ2AG6M:7443/arcgis/sharing/rest/generateToken redirects to https://EC2AMAZ-RQ2AG6M.local:7443/arcgis/sharing/rest/generateToken, and it's also worth noting that the common name for the self-signed certificate assigned to Portal for ArcGIS' web server is EC2AMAZ-RQ2AG6M.local.

Thanks again,
Trae

**Comment by traetimmerman** [comment]
Hey @cameronkroeker ,

I did some digging into the ArcGIS_Federation.psm1 file. It looks like the assignment of the PortalFQDN variable was the culprit. Since it references the PortalHostName variable during assignment. The PortalHostName variable was already assigned to the FQDN. Calling Get-FQDN on EC2AMAZ-RQ2AG6M.local returns EC2AMAZ-RQ2AG6M. I modified lines 133 and 340 to `$PortalFQDN = $PortalHostName` and reran InstallLicenseConfigure; federation was successful this time.

Thanks,
Trae

**Comment by cameronkroeker** [comment, marked as answer]
Thanks @traetimmerman, and great find. Your assessment is correct, it's already assigning the FQDN here:

https://github.com/Esri/arcgis-powershell-dsc/blob/3d056fe815c68e9676789afe9d4ca71c05cefc8d/Modules/ArcGIS/ArcGIS.psm1#L2060

Which means the FQDN was already getting passed into here:

https://github.com/Esri/arcgis-powershell-dsc/blob/3d056fe815c68e9676789afe9d4ca71c05cefc8d/Modules/ArcGIS/DSCResources/ArcGIS_Federation/ArcGIS_Federation.psm1#L133

The Get-FQDN only needs to be performed once. I suggest removing it from here though:

https://github.com/Esri/arcgis-powershell-dsc/blob/3d056fe815c68e9676789afe9d4ca71c05cefc8d/Modules/ArcGIS/ArcGIS.psm1#L2060

```
else{ $PrimaryPortalMachine.NodeName }
```

At any rate great find, and glad you found a solution that worked. We will look to get this fixed in a future release of the module. 

Thanks,
Cameron K.

**Reply by traetimmerman** [reply]
Thanks, @cameronkroeker. You mentioned earlier that it's best practice to use IP addresses when deploying in AWS. Would you be able to elaborate on why or point me toward some supporting documentation?

**Reply by cameronkroeker** [reply]
I'm not sure if there is supporting documentation, however I know the ArcGIS Enterprise Cloud Builder for AWS, and CloudFormation templates use the private ip addresses rather than the hostname. This is to avoid certain situations where the hostname "EC2AMAZ" may get assigned to more than one machine which would cause join-site failures in ArcGIS Server multi-machine configurations or highly available ArcGIS Enterprise.

**Reply by traetimmerman** [reply]
I see. I chose to go with the hostname to avoid issues with portal should I have to restore from an AMI. Thanks for the feedback.

**Comment by traetimmermantimmons** [comment]
Hey @cameronkroeker,

I've run into another use case that results in this issue. The deployment I'm working with is on-premises, and according to the Get-FQDN command in PowerShell, it looks like the FQDNs use mixed casing (e.g., server.Domain.Local). Token generation fails on the federation step, and I'm assuming this is because of the casing. Hitting the portal site using mixed casing redirects to site using all lower casing.

Just wanted to bring this to your attention in case there's a way to normalize casing in the logic in the federation step.

Thanks!
